### PR TITLE
Verify lower level transport exceptions don't bubble up on disconnects

### PR DESCRIPTION
**Pull request by bleskes**
#19096 introduced a generic TCPTransport base class so we can have multiple TCP based transport implementation. These implementations can vary in how they respond internally to situations where we concurrently send, receive and handle disconnects and can have different exceptions. However, disconnects are important events for the rest of the code base and should be distinguished from other errors (for example, it signals TransportMasterAction that it needs to retry and wait for the a (new) master to come back).  Therefore, we should make sure that all the implementations do the proper translation from their internal exceptions into ConnectTransportException which is used externally. 

Similarly we should make sure that the transport implementation properly recognize errors that were caused by a disconnect as such and deal with them correctly. This was, for example, the source of a build failure at https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-intake/1080 , where a concurrency issue cause SocketException to bubble out of MockTcpTransport.

This PR adds a tests which concurrently simulates connects, disconnects, sending and receiving and makes sure the above holds. It also fixes anything (not much!) that was found it.

### CODE REVIEW

**Comment by bleskes** [comment]
@s1monw can you take a careful look? I'm new to this code.

File: core/src/main/java/org/elasticsearch/common/transport/NetworkExceptionHelper.java [hunk 1]
```diff
@@ -35,7 +36,7 @@ public static boolean isConnectException(Throwable e) {
     }
 
     public static boolean isCloseConnectionException(Throwable e) {
-        if (e instanceof ClosedChannelException) {
+        if (e instanceof ClosedChannelException || e instanceof SocketException) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/common/transport/NetworkExceptionHelper.java]: I don't think that is true, SocketException is used in many different ways in the jdk we shouldn't generalize it in such a way

File: core/src/main/java/org/elasticsearch/common/transport/NetworkExceptionHelper.java [hunk 2]
```diff
@@ -55,6 +56,9 @@ public static boolean isCloseConnectionException(Throwable e) {
             if (e.getMessage().contains("Connection timed out")) {
                 return true;
             }
+            if (e.getMessage().contains("Socket is closed")) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/common/transport/NetworkExceptionHelper.java]: ++ maybe even stricter s/contains/equals?

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java [hunk 3]
```diff
@@ -997,9 +992,14 @@ public void sendResponse(Version nodeVersion, Channel channel, final TransportRe
                     transportServiceAdapter.onResponseSent(requestId, action, response, finalOptions);
                 }
             };
-            sendMessage(channel, reference, onRequestSent, false);
+            try {
+                sendMessage(channel, reference, onRequestSent, false);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: I wonder if we should have an internal private method that does this for use like:
`addedReleaseListener = sendMessageInternal(...)` so we do this only once?

File: core/src/main/java/org/elasticsearch/transport/local/LocalTransportChannel.java [hunk 4]
```diff
@@ -107,12 +107,7 @@ public void sendResponse(Exception exception) throws IOException {
 
     private void sendResponseData(byte[] data) {
         close();
-        targetTransport.workers().execute(() -> {
-            ThreadContext threadContext = targetTransport.threadPool.getThreadContext();
-            try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
-                targetTransport.messageReceived(data, action, sourceTransport, version, null);
-            }
-        });
+        targetTransport.receiveMessage(version, data, action, null, sourceTransport);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/transport/local/LocalTransportChannel.java]: ++

File: test/framework/src/main/java/org/elasticsearch/transport/AbstractSimpleTransportTestCase.java [hunk 5]
```diff
@@ -483,6 +488,122 @@ public void onNodeDisconnected(DiscoveryNode node) {
         assertThat(latch.await(5, TimeUnit.SECONDS), equalTo(true));
     }
 
+    public void testConcurrentSendRespondAndDisconnect() throws BrokenBarrierException, InterruptedException {
```
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/AbstractSimpleTransportTestCase.java]: ++ crazy but ok :)

**Comment by s1monw** [comment]
left some minors LGTM in general

File: test/framework/src/main/java/org/elasticsearch/transport/AbstractSimpleTransportTestCase.java (hunk 5 above)
Comment (bleskes) [inline test/framework/src/main/java/org/elasticsearch/transport/AbstractSimpleTransportTestCase.java]: yeah, it took some effort to reproduce the original issue :)

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java (hunk 3 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: yeah can do.

File: core/src/main/java/org/elasticsearch/common/transport/NetworkExceptionHelper.java (hunk 2 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/common/transport/NetworkExceptionHelper.java]: sure, I followed the same pattern as the other ifs.

File: core/src/main/java/org/elasticsearch/common/transport/NetworkExceptionHelper.java (hunk 1 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/common/transport/NetworkExceptionHelper.java]: out of curiosity - what other SocketExceptions should be treated as the if the connection is closed (i.e., it should be closed if it isn't)?

Comment (s1monw) [inline core/src/main/java/org/elasticsearch/common/transport/NetworkExceptionHelper.java]: just look at where that ctor is used there are so many instance of this I think just assuming that the socket is closed is wrong. one way of doing it is to check the closed flag on the channel and throw ChannelClosedExcepiton and add the SocketException as a suppressed one

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/common/transport/NetworkExceptionHelper.java]: >  the closed flag on the channel 

What flag do you mean exactly?

> just look at where that ctor is used there are so many instance of this I think just assuming that the socket is closed is wrong

OK. I'll remove the check. We can always reduce noise if we see too many warn logs.

**Comment by bleskes** [comment]
@s1monw thx for the feedback. I updated the PR

**Comment by s1monw** [comment]
LGTM

**Pull request merged**